File: python/391_PerfectRectangle.py
from unittest import TestCase
# https://leetcode.com/problems/perfect-rectangle/


class PerfectRectangle(TestCase):
    def isRectangleCover(self, rectangles):
        """
        :type rectangles: List[List[int]]
        :rtype: bool
        """
        def area(lx, ly, rx, ry):
            return (ry - ly)*(rx-lx)

        def corners(lx, ly, rx, ry):
            return {(lx, ly), (lx, ry), (rx, ly), (rx, ry)}

        sum = 0
        xor = set()
        for r in rectangles:
            sum += area(*r)
            xor ^= corners(*r)
        a, b, c, d = zip(*rectangles)
        whole = (min(a), min(b), max(c), max(d))
        return sum == area(*whole) and xor == corners(*whole)

    def test1(self):
        self.assertEqual(True, self.isRectangleCover([
            [1,1,3,3],
            [3,1,4,2],
            [3,2,4,4],
            [1,3,2,4],
            [2,3,3,4]
        ]))

    def test2(self):
        self.assertEqual(False, self.isRectangleCover([
            [1, 1, 3, 3],
            [3, 1, 4, 2],
            [1, 3, 2, 4],
            [3, 2, 4, 4]
        ]))

    def test3(self):
        self.assertEqual(False, self.isRectangleCover([
            [1, 1, 3, 3],
            [3, 1, 4, 2],
            [1, 3, 2, 4],
            [2, 2, 4, 4]
        ]))

File: kotlin/src/main/kotlin/391.PerfectRectangle.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/perfect-rectangle/


class PerfectRectangle{
  fun isRectangleCover(rectangles: Array<IntArray>): Boolean {
    var (glx, gly, grx, gry) = intArrayOf(Int.MAX_VALUE, Int.MAX_VALUE, 0, 0)
    var sum = 0
    val points = HashSet<Pair<Int, Int>>()

    fun merge(rect: IntArray){
      val (lx, ly, rx, ry) = rect
      for(point in listOf(lx to ly, lx to ry, rx to ly, rx to ry)){
        if(points.contains(point))
          points.remove(point)
        else
          points.add(point)
      }
    }
    for(r in rectangles){
      val (lx, ly, rx, ry) = r
      sum += (rx-lx)*(ry-ly)
      merge(r)
      glx = minOf(glx, lx)
      gly = minOf(gly, ly)
      grx = maxOf(grx, rx)
      gry = maxOf(gry, ry)
    }
    merge(intArrayOf(glx, gly, grx, gry))
    return points.isEmpty() && sum == (gry-gly)*(grx-glx)
  }

  @Test
  fun test1(){
    assertEquals(true, isRectangleCover(arrayOf(
        intArrayOf(1,1,3,3),
        intArrayOf(3,1,4,2),
        intArrayOf(3,2,4,4),
        intArrayOf(1,3,2,4),
        intArrayOf(2,3,3,4)
    )))
  }

  @Test
  fun test2(){
    assertEquals(false, isRectangleCover(arrayOf(
        intArrayOf(1,1,2,3),
        intArrayOf(3,1,4,2),
        intArrayOf(1,3,2,4),
        intArrayOf(3,2,4,4)
    )))
  }

  @Test
  fun test3(){
    assertEquals(false, isRectangleCover(arrayOf(
        intArrayOf(1,1,3,3),
        intArrayOf(3,1,4,2),
        intArrayOf(1,3,2,4),
        intArrayOf(2,2,4,4)
    )))
  }
}